### DISCUSSION TITLE
Able to install a custom .whl package in the extended image, but MethodNotFound from the DAG that uses it.

### DISCUSSION THREAD

**Discussion by rapala61** [apache/airflow#36030] [General] [answered]
Hi everyone, I've been building an SFTP -> Snowflake platform for the past quarter. 

Airflow: 2.7.2
Python: 3.11.4
Constrains: `--constraint "https://raw.githubusercontent.com/apache/airflow/constraints-2.7.2/constraints-3.11.txt"`

Right now, the problem I have is that I can install the `.whl` package in the extended Dockerfile, but Airflow complains of the package not being available for import.

I feel this is a product of me not fully understanding under what user is the package being installed and what user is being used when generating the DAGs.

Any idea where to look or what may be the problem is welcome. 🙂 

<img width="1504" alt="Screenshot 2023-12-02 at 6 58 52 PM" src="https://github.com/apache/airflow/assets/863301/864364b7-ea61-4a0a-8ef5-73cf0070b869">

<img width="611" alt="Screenshot 2023-12-02 at 6 59 11 PM" src="https://github.com/apache/airflow/assets/863301/e215faa3-1520-4010-ab75-878ece4fd906">

<img width="687" alt="Screenshot 2023-12-02 at 7 01 55 PM" src="https://github.com/apache/airflow/assets/863301/e4190365-fe7b-4779-9c54-0f933241f38d">

**Comment by potiuk** [comment, marked as answer]
You seem to have double underscore before `jw`. Hard to say if that's the issue. also hard to say about installation - because you  have not shared your dockerfile.

Generally all examples, how to build image with many examples is in https://airflow.apache.org/docs/docker-stack/index.html - when you are using airflow image. But It's hard to say what you use.

There are plenty of tools you can use to check if your image contains installed packages `pip freeze`, `pipdeptree`, `python -m site` to check where your installation is etc. so you can inspect those. Without you checking what it's there and inspecting things, it's really a guesswork.

**Reply by rapala61** [reply]
Hey, thanks for the response @potiuk . It was a combination of problems. his is how I checked if the module was available or not: `docker exec jw-data-sftp-report-ingestion-platform-airflow-scheduler-1 python -c "import sys; print(sys.path); print(len(sys.path)); from sftp_config_parser_jw.config_parser import partner_configs;"` It's working now. I did refactor that extra `_` before "jw". I'm sure it was one of the main issues. 

Posting my Dockerfile in case is helpful for context.

```
FROM apache/airflow:2.7.2-python3.11
USER root
RUN apt-get update \
  && apt-get install -y --no-install-recommends \
         vim build-essential libpq-dev tree\
  && apt-get autoremove -yqq --purge \
  && apt-get clean \
  && rm -rf /var/lib/apt/lists/*
WORKDIR /opt/airflow
USER airflow
ADD local-dev-airflow-requirements.txt .
RUN pip install --no-cache-dir --upgrade pip \
  && pip install --no-cache-dir -r local-dev-airflow-requirements.txt \
  pip install --no-cache-dir psycopg2==2.9.9
# Installing custom python packages
# Add config_parser whl
ADD config_parser/dist/sftp_config_parser_jw-*.whl wheels/
RUN --network=none pip install --no-index --find-links wheels sftp_config_parser_jw
```
